test/unit/test_data_utils.py

```diff
@@ -225,3 +226,19 @@ def test_parse_protobuf_dmatrix_single_feature_label(self):
                 pb_path = os.path.join(self.data_path, 'recordio_protobuf', file_path)
                 reader = data_utils.get_recordio_protobuf_dmatrix
                 self._check_dmatrix(reader, pb_path, 1, 1)
+
+    @patch("logging.warning")
+    def test_check_data_redundancy_positive(self, mock_log_warning):
```

Comment:
Let's add additional test(s) to cover file and pipe mode differences since that failed tests last time.